tests/check_framework/test_templates.py

```diff
@@ -1,162 +1,161 @@
 
-    @override_settings(TEMPLATES=TEMPLATES_APP_DIRS_AND_LOADERS)
-    def test_app_dirs_and_loaders(self):
-        """
-        Error if template loaders are specified and APP_DIRS is True.
-        """
-        self.assertEqual(check_setting_app_dirs_loaders(None), [E001])
+    def check(self, **kwargs):
+        return [Error("Example")]
 
-    def test_app_dirs_removed(self):
-        TEMPLATES = deepcopy(self.TEMPLATES_APP_DIRS_AND_LOADERS)
-        del TEMPLATES[0]["APP_DIRS"]
-        with self.settings(TEMPLATES=TEMPLATES):
-            self.assertEqual(check_setting_app_dirs_loaders(None), [])
 
-    def test_loaders_removed(self):
-        TEMPLATES = deepcopy(self.TEMPLATES_APP_DIRS_AND_LOADERS)
-        del TEMPLATES[0]["OPTIONS"]["loaders"]
-        with self.settings(TEMPLATES=TEMPLATES):
-            self.assertEqual(check_setting_app_dirs_loaders(None), [])
+class CheckTemplatesTests(SimpleTestCase):
+    @override_settings(
+        TEMPLATES=[
+            {"BACKEND": f"{__name__}.{ErrorEngine.__qualname__}", "NAME": "backend_1"},
+            {"BACKEND": f"{__name__}.{ErrorEngine.__qualname__}", "NAME": "backend_2"},
+        ]
+    )
+    def test_errors_aggregated(self):
+        errors = check_templates(None)
+        self.assertEqual(errors, [Error("Example")] * 2)
 
 
 class CheckTemplateStringIfInvalidTest(SimpleTestCase):
     TEMPLATES_STRING_IF_INVALID = [
         {
             "BACKEND": "django.template.backends.django.DjangoTemplates",
+            "NAME": "backend_1",
             "OPTIONS": {
                 "string_if_invalid": False,
             },
         },
         {
             "BACKEND": "django.template.backends.django.DjangoTemplates",
+            "NAME": "backend_2",
             "OPTIONS": {
                 "string_if_invalid": 42,
             },
         },
     ]
 
-    @classmethod
-    def setUpClass(cls):
-        super().setUpClass()
-        cls.error1 = copy(E002)
-        cls.error2 = copy(E002)
-        string_if_invalid1 = cls.TEMPLATES_STRING_IF_INVALID[0]["OPTIONS"][
-            "string_if_invalid"
-        ]
-        string_if_invalid2 = cls.TEMPLATES_STRING_IF_INVALID[1]["OPTIONS"][
-            "string_if_invalid"
-        ]
-        cls.error1.msg = cls.error1.msg.format(
-            string_if_invalid1, type(string_if_invalid1).__name__
+    def _get_error_for_engine(self, engine):
+        value = engine.engine.string_if_invalid
+        return Error(
+            f"'string_if_invalid' in TEMPLATES OPTIONS must be a string "
+            f"but got: {value!r} ({type(value)}).",
+            obj=engine,
+            id="templates.E002",
         )
-        cls.error2.msg = cls.error2.msg.format(
-            string_if_invalid2, type(string_if_invalid2).__name__
+
+    def _check_all_engines(self, all_engines):
+        return list(
+            chain.from_iterable(
+                e._check_string_if_invalid_is_string() for e in all_engines
+            )
         )
```

Comment:
This no longer checks *all* engines, but the passed in engines. Can you change the naming accordingly? And in tests below.
```suggestion
    def _check_engines(self, engines):
        return list(
            chain.from_iterable(
                e._check_string_if_invalid_is_string() for e in engines
            )
        )
```